Meson buildsystem support

eessmann opened the pull request:
This pull request introduces the Meson build system as an alternative to the existing Makefile-based system for building the rcm-f90 library. Meson offers several advantages in terms of robustness, cross-platform compatibility, and ease of use.

pbartholomew08 commented:
From past experience having multiple build systems is not a good idea. If Makefiles are not supported on other platforms (Windows?) then I think the better solution is to use something like CMake to generate a build system.

pbartholomew08 commented:
To discuss: meson vs cmake

eessmann commented:
Converted it to use CMake. Removed make and meson.

pbartholomew08 commented on .gitignore:
This seems excessive... Can you reduce it back to the original and only add what is required? For instance I don't think we need to worry about SVN

pbartholomew08 commented on README.md:
Please restore

pbartholomew08 replied:
Actually ignore this

pbartholomew08 commented on README.rst:
Please delete, and move any changes into README.md

pbartholomew08 replied:
Actually, igore this. No point in duplicate work

pbartholomew08 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+# Set up install directories
+include(GNUInstallDirs)
+set(INSTALL_INCLUDEDIR "${CMAKE_INSTALL_INCLUDEDIR}/rcm-f90")
+set(INSTALL_MODULEDIR "${INSTALL_INCLUDEDIR}/mod")
```
The current Makefile uses `include/` for module files, to avoid changing users we should keep the same directory layout.

pbartholomew08 commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+add_library(rcm-f90 ${rcm-sources})
+
+# Store generated mod-files in a separate folder
+set(moddir "${CMAKE_CURRENT_BINARY_DIR}/mod")
```
Is this not set by the parent CMakeLists.txt? Also, as there please use `include/` to avoid changing other projects.

pbartholomew08 commented on CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+)
+
+# Set up install directories
+include(GNUInstallDirs)
+set(INSTALL_INCLUDEDIR "${CMAKE_INSTALL_INCLUDEDIR}/rcm-f90")
+set(INSTALL_MODULEDIR "${INSTALL_INCLUDEDIR}/mod")
```
Similarly to the `include/` comments, please install the library under `lib/`.

As discussed yesterday there is some desire to have static/shared library functionality - see https://cmake.org/cmake/help/latest/guide/tutorial/Selecting%20Static%20or%20Shared%20Libraries.html - no strong opinions on the default. The link also details some library path settings.

pbartholomew08 replied:
The GNUInstallDirs module sets the library installation path to `lib/`, no need to set explicitly

eessmann replied:
Yup, the GNUInstallDirs should set the install paths for lib and include to their normal paths. I'll make sure that everything is install in the same places as the make version

eessmann commented:
Hi Paul,
I implemented the changes you suggested. Additionally I switched the flags to use an interface target and prepended a rcm to prevent any of rcm flags or options lacking out if its consumed downstream as a sub-project. Also I added flags for the cray and cray-llvm compiler, I'll add the rest of the compilers asimov regularly uses in the future.

For neatness I also converted the readme back to rst and edited it a bit
